azurerm_postgresql_server public_network_access_enabled setting on replica not respected on create

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and AzureRM Provider) Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->
```
Terraform v0.14.9
+ provider registry.terraform.io/hashicorp/azurerm v2.55.0

Your version of Terraform is out of date! The latest version
is 0.14.10. You can update by downloading from https://www.terraform.io/downloads.html
```
### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `azurerm_postgresql_server`

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Two files:

**postgres.tf**
```hcl
resource "azurerm_resource_group" "rg" {
  name     = var.resource_group
  location = var.main_location
}


resource "azurerm_postgresql_server" "postgres_master" {
  name                             = var.main_db_name
  location                         = var.main_location
  resource_group_name              = azurerm_resource_group.rg.name
  sku_name                         = var.sku_name
  administrator_login              = var.administrator_login
  administrator_login_password     = var.administrator_login_password
  version                          = var.postgres_version
  ssl_enforcement_enabled          = true
  ssl_minimal_tls_version_enforced = "TLS1_2"
  backup_retention_days            = var.retention_days
  geo_redundant_backup_enabled     = var.replicas_count != 0 ? true : false
  storage_mb                       = var.storage_mb
  public_network_access_enabled    = var.public_access_enabled
  auto_grow_enabled                = true

  lifecycle {
    ignore_changes = [
      # Autogrow is enabled
      storage_mb,
    ]
  }

}

resource "azurerm_postgresql_server" "postgres_standby" {
  count                            = var.replicas_count
  name                             = "${azurerm_postgresql_server.postgres_master.name}-r-${var.replicas_count}"
  location                         = var.replicas_location
  resource_group_name              = var.resource_group
  sku_name                         = var.sku_name
  version                          = var.postgres_version
  ssl_enforcement_enabled          = true
  ssl_minimal_tls_version_enforced = "TLS1_2"
  storage_mb                       = var.storage_mb
  public_network_access_enabled    = var.public_access_enabled
  create_mode                      = "Replica"
  creation_source_server_id        = azurerm_postgresql_server.postgres_master.id
  auto_grow_enabled                = true

  lifecycle {
    ignore_changes = [
      # Autogrow is enabled
      storage_mb,
    ]
  }

}
```
**var.tf**
```hcl
provider "azurerm" {
  features {}
}

variable "main_db_name" {
    default = "nfpgtst"
}

variable "main_location" {
    default = "westus2"
}
variable "resource_group" {
    default = "pg-test-diff-reg"
}
variable "sku_name" {
    default = "GP_Gen5_4"
}
variable "administrator_login" {
    default = "nilfranadmin"
}
variable "administrator_login_password" {
    default = "superSecure123$"
}
variable "postgres_version" {
    default = 9.6
}
variable "retention_days" {
    default = 7
}
variable "replicas_count" {
    default = 1
}
variable "storage_mb" {
    default = 5120
}
variable "replicas_location" {
    default = "eastus2"
}
variable "public_access_enabled" {
    default = false
    }

```
### Debug Output
https://gist.github.com/NillsF/be2b1d564fa20844f7e59a98c8dc8379
<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behaviour
- Replica's `public_network_access_enabled` should be kept false when creating.
### Actual Behaviour
When creating a postgresSQL replica, the `public_network_access_enabled` setting is set to true. In the master, it is kept the right way.

However, the `terraform plan` shows TF trying to set it to false:
```
# azurerm_postgresql_server.postgres_standby[0] will be created
  + resource "azurerm_postgresql_server" "postgres_standby" {
      + administrator_login              = (known after apply)
      + auto_grow_enabled                = true
      + backup_retention_days            = (known after apply)
      + create_mode                      = "Replica"
      + creation_source_server_id        = (known after apply)
      + fqdn                             = (known after apply)
      + geo_redundant_backup_enabled     = (known after apply)
      + id                               = (known after apply)
      + location                         = "westus2"
      + name                             = "nfpgsame-r-1"
      + public_network_access_enabled    = false
      + resource_group_name              = "pg-test-same-reg"
      + sku_name                         = "GP_Gen5_2"
      + ssl_enforcement                  = (known after apply)
      + ssl_enforcement_enabled          = true
      + ssl_minimal_tls_version_enforced = "TLS1_2"
      + storage_mb                       = 5120
      + version                          = "9.6"

      + storage_profile {
          + auto_grow             = (known after apply)
          + backup_retention_days = (known after apply)
          + geo_redundant_backup  = (known after apply)
          + storage_mb            = (known after apply)
        }
    }

```

This can be resolved by running `terraform plan` and `terraform apply` a second time.
<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->
1. Run `terraform plan` and `terraform apply` using the files provided above. This will result in the bad config
2. Running `terraform plan` and `terraform apply` again will update the configuration correctly.

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

@aristosvo did some great work implementing locking in the following issue/PR:
* #11318
* #11330

It's a tough problem, although not hard to fix. I believe it is a problem which should probably be fixed on the Azure API side.

I've reproduced it in the acctests by adding `public_network_access_enabled = true` to the replica in ` TestAccPostgreSQLServer_createReplica`, resulting in a test failure:
```
❯ make acctests SERVICE='postgres' TESTARGS='-run=createReplica'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/postgres -run=createReplica -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/04/24 15:29:08 [DEBUG] not using binary driver name, it's no longer needed
2021/04/24 15:29:09 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccPostgreSQLServer_createReplica
=== PAUSE TestAccPostgreSQLServer_createReplica
=== CONT  TestAccPostgreSQLServer_createReplica
    testing.go:620: Step 3/4 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_postgresql_server.replica will be updated in-place
          ~ resource "azurerm_postgresql_server" "replica" {
                id                                = "/subscriptions/6eaaf383-01c1-4431-921c-73577a49128d/resourceGroups/acctestRG-psql-210424152910410484/providers/Microsoft.DBforPostgreSQL/servers/acctest-psql-server-210424152910410484-replica"
                name                              = "acctest-psql-server-210424152910410484-replica"
              ~ public_network_access_enabled     = true -> false
                # (16 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccPostgreSQLServer_createReplica (665.28s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres  668.338s
FAIL
make: *** [acctests] Error 1
```

--- a/azurerm/internal/services/postgres/postgresql_server_resource.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource.go
@@ -609,7 +609,7 @@ func resourcePostgreSQLServerCreate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
-	// Issue tracking REST API update: https://github.com/Azure/azure-rest-api-specs/issues/14117
+	// Issue tracking the REST API update failure: https://github.com/Azure/azure-rest-api-specs/issues/14117
 	if mode == postgresql.CreateModeReplica {
 		log.Printf("[INFO] changing `public_network_access_enabled` for AzureRM PostgreSQL Server %q (Resource Group %q)", name, resourceGroup)
 		properties := postgresql.ServerUpdateParameters{

--- a/azurerm/internal/services/postgres/postgresql_server_resource.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource.go
@@ -609,6 +609,25 @@ func resourcePostgreSQLServerCreate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
+	// Issue tracking REST API update: https://github.com/Azure/azure-rest-api-specs/issues/14117
+	if mode == postgresql.CreateModeReplica {
+		log.Printf("[INFO] changing `public_network_access_enabled` for AzureRM PostgreSQL Server %q (Resource Group %q)", name, resourceGroup)
+		properties := postgresql.ServerUpdateParameters{
+			ServerUpdateParametersProperties: &postgresql.ServerUpdateParametersProperties{
+				PublicNetworkAccess: publicAccess,
+			},
+		}
+
+		future, err := client.Update(ctx, resourceGroup, name, properties)
+		if err != nil {
+			return fmt.Errorf("updating PostgreSQL Server %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+
+		if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+			return fmt.Errorf("waiting for update of PostgreSQL Server %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+
 	return resourcePostgreSQLServerRead(d, meta)
 }
 

--- a/azurerm/internal/services/postgres/postgresql_server_resource_test.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource_test.go
@@ -768,7 +768,8 @@ resource "azurerm_postgresql_server" "replica" {
   create_mode               = "Replica"
   creation_source_server_id = azurerm_postgresql_server.test.id
 
-  ssl_enforcement_enabled = true
+  public_network_access_enabled = false
+  ssl_enforcement_enabled       = true
 }
 `, r.template(data, sku, "11"), data.RandomInteger, data.Locations.Secondary, sku)
 }

